test/puppetlabs/puppetdb/generative/submit_command.clj

```diff
@@ -0,0 +1,150 @@
+                :puppet/producer_timestamp timestamp-gen
+                :command/received timestamp-gen
+                :puppet/producer (gen/elements ["a.com" "b.com" "c.com"])
+                :catalog/resource-parameters (gen/map (pdb-gen/string+)
+                                                      (gen/one-of [(pdb-gen/string+)
+                                                                   pdb-gen/pg-smallint
+                                                                   gen/boolean])
+                                                      {:min-elements 0
+                                                       :max-elements 3})})
+
+;;; Utilities for comparing puppetdb states
+
+(defn all-catalogs [db]
+  (->> (qeng/stream-query-result :v4
+                             ["from" "catalogs"]
+                             {:order_by [[:certname :descending]
+                                         [:producer_timestamp :descending]]}
+                             {:scf-read-db db, :url-prefix "/pdb"})
+       (map (fn [row]
+              (-> row
+                  (update :resources #(json/parse (str % ) true))
+                  (update :edges #(json/parse (str % ) true))
+                  (update-in [:resources :data] set))))))
+
+(defn all-factsets [db]
+  (->> (qeng/stream-query-result :v4
+                                 ["from" "factsets"]
+                                 {:order_by [[:certname :descending]
+                                             [:producer_timestamp :descending]]}
+                                 {:scf-read-db db, :url-prefix "/pdb"})
+       (map (fn [row]
+              (-> row
+                  (update :facts #(json/parse (str % ) true))
+                  (update-in [:facts :data] #(sort-by :name %)))))))
+
+(defn all-reports [db]
+  (->> (qeng/stream-query-result :v4
+                                 ["from" "reports"]
+                                 {:order_by [[:certname :descending]
+                                             [:producer_timestamp :descending]]}
+                                 {:scf-read-db db, :url-prefix "/pdb"})
+       (map (fn [row]
+              (-> row
+                  (update :logs #(json/parse (str %) true))
+                  (update :metrics #(json/parse (str %) true))
+                  (update :resource_events #(json/parse (str %) true)))))))
+
+(defn all-nodes [db]
+  (qeng/stream-query-result :v4
+                            ["from" "nodes"]
+                            {:order_by [[:certname :descending]]}
+                            {:scf-read-db db, :url-prefix "/pdb"}) )
+
+(defn state-snapshot [db]
+  {:nodes (all-nodes db)
+   :catalogs (all-catalogs db)
+   :factsets (all-factsets db)
+   :reports (all-reports db)})
+
+;;; Generative tests
+
+(tc/defspec submit-commands 25
+  (prop/for-all
+   [cmd (->> :puppetdb/command
+             (gen/override overrides)
+             gen/no-shrink
+             gen/convert)]
+   (print "-") (flush)
```

Comment:
I think this might be the first time we're intentionally printing things in a test. It seems useful when working on the tests locally, but it seems like it would just be noise in CI. Not sure what the best way to accommodate both is. Maybe log/trace or something? Thoughts @rbrw?